taichi/ir/frontend_ir.cpp

```diff
@@ -192,17 +215,53 @@ void BinaryOpExpression::type_check(CompileConfig *config) {
 
   if ((lhs_type->is<PrimitiveType>() && rhs_type->is<TensorType>()) ||
       (lhs_type->is<TensorType>() && rhs_type->is<PrimitiveType>())) {
-    TI_NOT_IMPLEMENTED;
+    // convert Tensor/Scalar | Scalar/Tensor operations to broadcasting
+    auto [unified_l, unified_r] = unify_binop_operands(lhs, rhs);
+    lhs = unified_l;
+    rhs = unified_r;
+    if (lhs->ret_type == PrimitiveType::unknown)
+      lhs.type_check(config);
+    if (rhs->ret_type == PrimitiveType::unknown)
+      rhs.type_check(config);
+    TI_ASSERT(lhs->ret_type->is<TensorType>());
```

Comment:
if `lhs_type->is<TensorType>() && rhs_type->is<PrimitiveType>()`, how come they get unified to TensorType?

Reply:
Unification here means converting to broadcast operations, so the `rhs` will be converted to a tensor of the previous `rhs` value that matches `lhs`'s shape.

Reply:
Oh I misunderstood this part, thanks!